### PR TITLE
Statement expressions

### PR DESCRIPTION
**Pull request by STRd6**
This is probably good enough to merge in.

- Removed IfExpression as separate part of grammar
  - Now get hoisted as StatementExpressions or wrapped in IIFEs similar to IterationExpression
- Made DeclarationConditions work with StatementExpressions
- Added more types
- Moved skipImplicitArguments check from parser to .civet file
- Moved declaration logic to parser/declaration.civet
- Handle simple assignments expressions

To add later:

- Extend wrap IIFE to handle return from outer function, yield.

Related: #380, #202

### CODE REVIEW

**Review by edemaine — APPROVED** [review]
This mostly seems like a great improvement: many IIFEs get removed.

The main regression is that `if` expressions aren't rewritten into equivalent `?:` when they're simple expressions like `x = if y then a else b`. This makes me sad, and performance-wise it's not ideal to add the extra `ref` variable that results. But it's also still correct, so we could leave improvement to the future.  (I haven't found any case that worked previously and doesn't work with this PR.)

I see two approaches for optimization:
* Detect expressionizable statements and more aggressively convert in this case. We already need this to detect IIFE wrapping (in a TODO you added), for cases like declarations inside `if` expressions.
* It looks like the currently generated code could be further simplified in many simple cases: an extra `ref` variable isn't always necessary. For example, we could do direct assignment instead of assigning to `ref` and then assigning to the desired variable. This would remove the need for `ref`, and remove the main inefficiency with this compilation.

I'm not sure I'd call #202 or #380 fully resolved; it's only resolved for top-level assignments and declarations. But it's a good start... And fully resolving will require a rather intimidating amount of code rewriting...

File: test/if.civet [hunk 1]
```diff
@@ -443,10 +443,12 @@ describe "if", ->
...
-      x = (y?
-        "a"
-      :
-        "b")
+      let ref;if (y) {
+        ref = "a"
+      }
+      else {
+        ref = "b"
+      };x = ref
```
[the remark this answers is not in the document]
Comment (STRd6) [reply test/if.civet:451]: We could avoid the intermediate in normal assignment but in `const` declarations we can't really. It's probably a special case we could optimize further as our internals get better typed and it is easier to shift things around.

File: test/if.civet [hunk 2]
```diff
@@ -625,9 +637,9 @@ describe "if", ->
       x = if (y)
         {}
       ---
-      x = (y?(
-        {}
-      ):void 0)
+      let ref;if (y) {
+        ref = ({})
+      } else {ref = undefined};x = ref
```
[the remark this answers is not in the document]
Comment (STRd6) [reply test/if.civet:642]: No particular reason, I think it's just hitting a different code path.

**Comment by edemaine** [comment]
By the way, `yarn build` gives a bunch of warnings because of mistaken `import` instead of `import type`; should probably be fixed.

**Comment by STRd6** [comment]
> This mostly seems like a great improvement: many IIFEs get removed.
> 
> The main regression is that `if` expressions aren't rewritten into equivalent `?:` when they're simple expressions like `x = if y then a else b`. This makes me sad, and performance-wise it's not ideal to add the extra `ref` variable that results. But it's also still correct, so we could leave improvement to the future. (I haven't found any case that worked previously and doesn't work with this PR.)
> 
> I see two approaches for optimization:
> 
> * Detect expressionizable statements and more aggressively convert in this case. We already need this to detect IIFE wrapping (in a TODO you added), for cases like declarations inside `if` expressions.

It should be possible to find simple ones and convert to ternary in those case.

> * It looks like the currently generated code could be further simplified in many simple cases: an extra `ref` variable isn't always necessary. For example, we could do direct assignment instead of assigning to `ref` and then assigning to the desired variable. This would remove the need for `ref`, and remove the main inefficiency with this compilation.

Also probably not too difficult, essentially pass the variable to be assigned as the ref rather than generating a fresh one.

We can address these in some follow up work.

> I'm not sure I'd call #202 or #380 fully resolved; it's only resolved for top-level assignments and declarations. But it's a good start... And fully resolving will require a rather intimidating amount of code rewriting...

True, it is more "resolved for normal cases" which is hopefully 95%+. To fully resolve we'd need to unwrap arbitrary expressions into a chain of statements. Not impossible but probably at least as much work as this was if not more. And, as a user, in almost every case it would probably be clearer to just rewrite the code in a simpler way to begin with.

**Comment by edemaine** [comment]
I think the latest commits might actually fully fix #380, at least barring more bugs in isExpression. #202 is the hard one.

**Pull request merged**